File: src/array/1806.cpp
#include <vector>
#include <string>
#include <algorithm>
#include "iostream"

using namespace std;

class Solution
{
public:
  int reinitializePermutation(int n)
  {
    int i, count;
    vector<int> arr, temp;

    count = 0;
    for (i = 0; i < n; i++)
    {
      arr.push_back(i);
      temp.push_back(i);
    }
    do
    {
      for (i = 0; i < n; i++)
      {
        if (i % 2)
        {
          temp[i] = arr[n / 2 + (i - 1) / 2];
        }
        else
        {
          temp[i] = arr[i / 2];
        }
      }
      arr = temp;
      count++;
    } while (!check(arr, n));
    return count;
  }

private:
  // 检测数字是否满足要求
  bool check(vector<int> arr, int n)
  {
    int i;
    for (i = 0; i < n; i++)
    {
      if (arr[i] != i)
      {
        return false;
      }
    }
    return true;
  }
};

int main()
{
  auto aa = new Solution();
  cout << aa->reinitializePermutation(4) << endl;
  return 0;
}

File: src/array/1806.js
/**
 * 判断数组中的下标是否与值相等
 * @param {number} arr
 */
const check = (arr) => {
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] !== i) {
      return false;
    }
  }
  return true;
};

/**
 * @param {number} n
 * @return {number}
 */
var reinitializePermutation = function (n) {
  let arr = [];
  let count = 0;
  for (let i = 0; i < n; i++) {
    arr.push(i);
  }

  do {
    const temp = [];

    for (let i = 0; i < n; i++) {
      const num = i % 2 ? arr[n / 2 + (i - 1) / 2] : arr[i / 2];
      temp[i] = num;
    }
    arr = temp;
    count++;
    // console.log(count, arr);
  } while (!check(arr) && count < 1000);
  return count;
};

for (let i = 2; i <= 100; i += 2) {
  console.log(i, reinitializePermutation(i));
}
